### ISSUE TITLE
Non-functioning gradient clipping repeatably leads to exploding gradients

### ISSUE THREAD

**Issue by amorehead** [Lightning-AI/pytorch-lightning#12595]
## 🐛 Bug

It looks like gradient clipping (both by norm and value with various threshold values) is not working in versions 1.5.10 and 1.6.0 of Lightning. If this observation is true, this may be the cause of exploding gradients in most of my recent trained models.

### To Reproduce

Train a LightningModule using `gradient_clip_val` in [0.0, 0.5, 1.0, 3.0, 10.0] with `grad_clip_algorithm` in ['value', 'norm'], and use a learning rate in [2e-7, 7e-7] (which equate to roughly 1e-3 when scaling them down by a factor of sqrt(num_effective_batch_size) = sqrt(48) in my case).

Example of exploding gradients in one of my models when applying gradient clipping with a value of 10.0 and using the norm algorithm:
![image](https://user-images.githubusercontent.com/7051982/161447403-cb57da04-d583-48bf-971e-8ac786c765e5.png)

The image above comes from logging gradients with WandB using `--track_grad_norm 2`.

### Expected behavior

I should be able to train a LightningModule with a reasonable learning rate, weight decay rate, and gradient clipping to avoid the exploding gradients problem.

### Additional context

Gradient clipping has been observed not to work on both an LSF server as well as a CentOS x86 server. In both settings, default FP32 and DDP was used.

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

**Comment by amorehead** [comment]
I can also confirm that the correct CLI arguments are being passed as I construct my Lightning `trainer` instances from argparse's namespaces: 
![image](https://user-images.githubusercontent.com/7051982/161449768-1183ea7f-39c4-499e-8342-44958dd1c959.png)

**Comment by amorehead** [comment]
Update: In my initial testing, I believe I've seen this issue disappear if I disable gradient accumulation completely (by setting it to one).

**Comment by rohitgr7** [comment]
noticed that gradients are logged before they are clipped. So during gradient accumulation, the gradients are accumulated and logged before they are clipped so that's why they are showing like this?

https://github.com/PyTorchLightning/pytorch-lightning/blob/f6d61cb08a9f9570cf8b2a11227bff4fa573fc31/pytorch_lightning/plugins/precision/precision_plugin.py#L117-L125

maybe we can move the logging after clipping? wdyt @carmocca ?

**Comment by carmocca** [comment]
I personally find logging before clipping more interesting assuming you trust clipping works. `track_grad_norm` is more of a debugging flag to me so I would rather know the previous value if I had to choose.

However, I understand the confusion. Maybe we want to log both with different suffixes

**Comment by magehrig** [comment]
If I understand correctly, the gradients are clipped? I would expect that gradients are clipped before logging by default

**Comment by carmocca** [comment]
> If I understand correctly, the gradients are clipped?

Yes.

Converting this to a feature request. The idea would be to log both with different suffixes

**Comment by schmittlema** [comment]
I'm having a similar or same issue where gradient clipping does not seem to be working properly.
I've set
```
gradient_clip_val=1.0,
gradient_clip_algorithm="norm",
```
but I'm seeing in wanb
```
wandb:         grad_2.0_norm/output_net.0.weight_epoch inf
wandb:                       grad_2.0_norm_total_epoch inf
```
Followed by -inf being output by the network. The issue above said that the values are recorded pre-clip, but I'm still seeing inf come out of the network suggesting otherwise

Is there something I'm missing here?

**Comment by amorehead** [comment]
I haven't seen this issue occur for me over the last year or so of using Lightning, so I'll close this issue for now.
